providerquerymanager/providerquerymanager.go

```diff
@@ -0,0 +1,377 @@
+		providerQueryMessages:        make(chan providerQueryMessage, 16),
+		providerRequestsProcessing:   make(chan cid.Cid),
+		incomingFindProviderRequests: make(chan cid.Cid),
+		inProgressRequestStatuses:    make(map[cid.Cid]*inProgressRequestStatus),
+		findProviderTimeout:          defaultTimeout,
+	}
+}
+
+// Startup starts processing for the ProviderQueryManager.
+func (pqm *ProviderQueryManager) Startup() {
+	go pqm.run()
+}
+
+type inProgressRequest struct {
+	providersSoFar []peer.ID
+	incoming       <-chan peer.ID
+}
+
+// SetFindProviderTimeout changes the timeout for finding providers
+func (pqm *ProviderQueryManager) SetFindProviderTimeout(findProviderTimeout time.Duration) {
+	pqm.timeoutMutex.Lock()
+	pqm.findProviderTimeout = findProviderTimeout
+	pqm.timeoutMutex.Unlock()
+}
+
+// FindProvidersAsync finds providers for the given block.
+func (pqm *ProviderQueryManager) FindProvidersAsync(sessionCtx context.Context, k cid.Cid, ses uint64) <-chan peer.ID {
+	inProgressRequestChan := make(chan inProgressRequest)
+
+	select {
+	case pqm.providerQueryMessages <- &newProvideQueryMessage{
+		ses:                   ses,
+		k:                     k,
+		inProgressRequestChan: inProgressRequestChan,
+	}:
+	case <-pqm.ctx.Done():
+		return nil
+	case <-sessionCtx.Done():
+		return nil
+	}
+
+	var receivedInProgressRequest inProgressRequest
+	select {
+	case <-sessionCtx.Done():
+		return nil
+	case receivedInProgressRequest = <-inProgressRequestChan:
+	}
+
+	return pqm.receiveProviders(sessionCtx, k, ses, receivedInProgressRequest)
+}
+
+func (pqm *ProviderQueryManager) receiveProviders(sessionCtx context.Context, k cid.Cid, ses uint64, receivedInProgressRequest inProgressRequest) <-chan peer.ID {
+	// maintains an unbuffered queue for incoming providers for given request for a given session
+	// essentially, as a provider comes in, for a given CID, we want to immediately broadcast to all
+	// sessions that queried that CID, without worrying about whether the client code is actually
+	// reading from the returned channel -- so that the broadcast never blocks
+	// based on: https://medium.com/capital-one-tech/building-an-unbounded-channel-in-go-789e175cd2cd
+	returnedProviders := make(chan peer.ID)
+	receivedProviders := append([]peer.ID(nil), receivedInProgressRequest.providersSoFar[0:]...)
+	incomingProviders := receivedInProgressRequest.incoming
+
+	go func() {
+		defer close(returnedProviders)
+		outgoingProviders := func() chan<- peer.ID {
+			if len(receivedProviders) == 0 {
+				return nil
+			}
+			return returnedProviders
+		}
+		nextProvider := func() peer.ID {
+			if len(receivedProviders) == 0 {
+				return ""
+			}
+			return receivedProviders[0]
+		}
+		for len(receivedProviders) > 0 || incomingProviders != nil {
+			select {
+			case <-sessionCtx.Done():
+				pqm.providerQueryMessages <- &cancelRequestMessage{
+					ses: ses,
+					k:   k,
+				}
+				// clear out any remaining providers
+				for range incomingProviders {
```

Comment:
Out of curiosity, why do this drain?

Reply:
don't want to block incase there are outstanding sends.

Reply:
but maybe there is a better way to do this.

Reply:
Thought about it a bit more, and yea, this is neccesary. The reason is this: 
messages get processed serially in the run loop. One of those messages might be an incoming provider found message that gets broadcast to everyone who is waiting. We don't want that to block, so we need to make sure the channel is read, emptied, and closed (which happens once the cancel is processed) before we proceed.